os/se/ameba/security_ameba_wrapper_tz.c

```diff
@@ -0,0 +1,826 @@
+
+/* Secure Storage Variable */
+#define SEC_STORE_OFFSET CONFIG_SE_SSTORAGE_ADDRESS
+#define SEC_STORE_MAX_SLOT 33
+
+/* Flash Sector Size */
+#define SECTOR_SIZE 4096
+
+/* Secure efuse key location */
+#define SAMSUNG_KEY_ADDR 0x150
+
+/* Non-Secure Data buff, 8K (2 Sector + Tag) */
+#define NS_BUF_LEN ((SECTOR_SIZE * 2) + 32)
+
+typedef struct {
+	void *ns_func_s;
+	uint8_t *buf;
+	uint32_t buf_len;
+} ns_passin_struc;
+
+typedef struct {
+	hal_data *input_data;
+	hal_data *output_data;
+} inout_struc;
+
+typedef struct {
+	uint32_t factorykey_device_idx;
+	uint32_t factorykey_cert_idx;
+	uint32_t factory_cert_addr;
+	uint32_t factory_key_addr;
+	uint32_t sstorage_addr;
+	uint32_t samsung_key_addr;
+} factory_struc;
+
+static uint8_t *ns_buf = NULL;
+
+extern void *rtl_set_ns_func(void);
+
+
+int se_ameba_hal_init(hal_init_param *params)
+{
+	AWRAP_ENTER;
+
+	int ret;
+	factory_struc input_data;
+	ns_passin_struc ns_passin;
+
+	/* Assign Pass in function and buff */
+	ns_buf = (uint8_t *)kmm_malloc(NS_BUF_LEN);
+	if (ns_buf == NULL) {
+		ret = HAL_NOT_ENOUGH_MEMORY;
+		goto exit;
+	}
+	ns_passin.ns_func_s = rtl_set_ns_func();
+	ns_passin.buf = ns_buf;
+	ns_passin.buf_len = NS_BUF_LEN;
+
+	/* Factory Key and Cert address (Flash) */
+	input_data.factorykey_device_idx = FACTORYKEY_RTL_DEVICE;
+	input_data.factorykey_cert_idx = FACTORYKEY_RTL_CERT;
+
+	/* Factory Key and Cert address (Flash) */
+	input_data.factory_cert_addr = FACTORY_CERT_ADDR;
+	input_data.factory_key_addr = FACTORY_KEY_ADDR;
+
+	/* Secure Storage Base address (Flash) */
+	input_data.sstorage_addr = SEC_STORE_OFFSET;
+	input_data.samsung_key_addr = SAMSUNG_KEY_ADDR;
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_init(params, &input_data, &ns_passin);
+	up_free_secure_context();
+
+exit:
+	if (ret != HAL_SUCCESS) {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_deinit(void)
+{
+	AWRAP_ENTER;
+	int ret;
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_deinit();
+	up_free_secure_context();
+
+	if (ret == HAL_SUCCESS) {
+		if (ns_buf != NULL) {
+			kmm_free(ns_buf);
+			ns_buf = NULL;
+		}
+	} else {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_free_data(hal_data *data)
+{
+	AWRAP_ENTER;
+	int ret;
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_free_data(data);
+	up_free_secure_context();
+
+	if (ret != HAL_SUCCESS) {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_get_status(void)
+{
+	AWRAP_ENTER;
+	int ret;
+
+	up_allocate_secure_context(4096);
+	ret = ameba_hal_get_status();
+	up_free_secure_context();
+
+	return ret;
+}
+
+int se_ameba_hal_set_key(hal_key_type mode, uint32_t key_idx, hal_data *key, hal_data *prikey)
+{
+	AWRAP_ENTER;
+	int ret;
+	inout_struc pub_prikey;
+
+	if (prikey != NULL) {	/* If private key is pass in */
+		return HAL_NOT_SUPPORTED;
+	}
+	if (key_idx >= KEY_STORAGE_INDEX_MAX) {	/* Index Out of Range */
```

Comment:
As my understanding, we can use index from 0 to MAX(32). If we put the `key_idx` to 32, then this treats as invalid.
So I think we should compare only with `>`, not `>=`.

Reply:
Total key slot is 32, so range is from 0 to 31.
Is my understanding correct? Please confirm.
Same for Secure Storage, total slot is 33. so range is 0 to 32.

Reply:
Oh, you are right. Let's keep this.